packages/timeline-state-resolver/src/devices/telemetrics.ts

```diff
@@ -0,0 +1,191 @@
+import * as _ from 'underscore'
+import { DoOnTime } from '../doOnTime'
+import Timer = NodeJS.Timer
+
+const TELEMETRICS_NAME = 'Telemetrics'
+const TELEMETRICS_COMMAND_PREFIX = 'P0C'
+const SOCKET_PORT = 5000
+const TIMEOUT_IN_MS = 2000
+
+interface TelemetricsState {
+	presetShotIdentifiers: number[]
+}
+
+/**
+ * Connects to a Telemetrics Device on port 5000 using a TCP socket.
+ * This class uses a fire and forget approach.
+ */
+export class TelemetricsDevice extends DeviceWithState<TelemetricsState, DeviceOptionsTelemetrics> {
+	private doOnTime: DoOnTime
+
+	private socket: Socket
+	private statusCode: StatusCode = StatusCode.UNKNOWN
+	private errorMessage: string
+	private resolveInitPromise: (value: boolean) => void
+
+	private retryConnectionTimer: Timer | undefined
+
+	constructor(deviceId: string, deviceOptions: DeviceOptionsTelemetrics, getCurrentTime: () => Promise<number>) {
+		super(deviceId, deviceOptions, getCurrentTime)
+
+		this.doOnTime = new DoOnTime(() => this.getCurrentTime())
+		this.handleDoOnTime(this.doOnTime, 'telemetrics')
+	}
+
+	get canConnect(): boolean {
+		return true
+	}
+
+	clearFuture(_clearAfterTime: number): void {
+		// No state to handle - we use a fire and forget approach
+	}
+
+	get connected(): boolean {
+		return this.statusCode === StatusCode.GOOD
+	}
+
+	get deviceName(): string {
+		return `${TELEMETRICS_NAME} ${this.deviceId}`
+	}
+
+	get deviceType() {
+		return DeviceType.TELEMETRICS
+	}
+
+	getStatus(): DeviceStatus {
+		const messages: string[] = []
+
+		switch (this.statusCode) {
+			case StatusCode.GOOD:
+				this.errorMessage = ''
+				messages.push('Connected')
+				break
+			case StatusCode.BAD:
+				messages.push('No connection')
+				break
+			case StatusCode.UNKNOWN:
+				this.errorMessage = ''
+				messages.push('Not initialized')
+				break
+		}
+
+		if (this.errorMessage) {
+			messages.push(this.errorMessage)
+		}
+
+		return {
+			statusCode: this.statusCode,
+			messages,
+			active: this.isActive,
+		}
+	}
+
+	handleState(newState: TimelineState, mappings: Mappings): void {
+		super.onHandleState(newState, mappings)
+		_.each(newState.layers, (timelineObject, _layerName) => {
+			const telemetricsObject: TimelineObjTelemetrics = timelineObject as unknown as TimelineObjTelemetrics
+
+			const state: { state: TelemetricsState; time: number } | null = this.getState(newState.time)
+			const telemetricsState: TelemetricsState = state === null ? { presetShotIdentifiers: [] } : state.state
+
+			telemetricsObject.content.presetShotIdentifiers.forEach((presetShotIdentifier: number) => {
+				if (telemetricsState.presetShotIdentifiers.includes(presetShotIdentifier)) {
+					return
+				}
+				telemetricsState.presetShotIdentifiers.push(presetShotIdentifier)
+				this.setState(telemetricsState, newState.time)
+				const command = `${TELEMETRICS_COMMAND_PREFIX}${presetShotIdentifier}\r`
+				this.doOnTime.queue(newState.time, undefined, () => this.socket.write(command))
+			})
+		})
```

Comment:
This should look somewhat like that
```suggestion
		const previousStateTime = Math.max(this.getCurrentTime(), newState.time)
		const oldState: TelemetricsState = this.getStateBefore(previousStateTime)?.state ?? { presetShotIdentifiers: [] }
		const newState: TelemetricsState = { presetShotIdentifiers: [] }

		_.each(newState.layers, (timelineObject, _layerName) => {
			const telemetricsObject: TimelineObjTelemetrics = timelineObject as unknown as TimelineObjTelemetrics
			telemetricsObject.content.presetShotIdentifiers.forEach((presetShotIdentifier: number) => {
				newState.presetShotIdentifiers.push(presetShotIdentifier)
			})
		})
		
		this._doOnTime.clearQueueNowAndAfter(previousStateTime)

		// diff the state here (queue commands only for preset identifiers from newState that don't exist in oldState)
		
		this.setState(newState, newState.time)
```

Reply:
I've gone away from using the `DeviceWithState` class so there is no longer anything to diff changes on :)